src/Components/RightClick/Base/RightClickMenu.tsx

```diff
@@ -0,0 +1,123 @@
+// Libraries
+import React, {forwardRef, RefObject, useLayoutEffect, useRef} from 'react'
+import classnames from 'classnames'
+
+// Components
+import {ClickOutside} from '../../ClickOutside/ClickOutside'
+
+// Utilities
+import {convertCSSPropertiesToString} from '../../../Utils/index'
+import {calculateRightClickMenuStyles} from '../../../Utils/rightClick'
+
+// Types
+import {
+  ComponentColor,
+  StandardFunctionProps,
+  Coordinates,
+} from '../../../Types'
+
+export interface RightClickMenuProps extends StandardFunctionProps {
+  /** Bounding rectangle of trigger element */
+  triggerRef: RefObject<any>
+  /** Menu dialog color */
+  color: ComponentColor
+  /** Dismisses the menu */
+  onHide: () => void
+  /** Mouse position from right click event */
+  mouseOffset: Coordinates
+}
+
+export type RightClickMenuRef = HTMLUListElement
+
+export const RightClickMenu = forwardRef<
+  RightClickMenuRef,
+  RightClickMenuProps
+>(
+  (
+    {
+      id,
+      style,
+      color,
+      onHide,
+      testID,
+      children,
+      className,
+      triggerRef,
+      mouseOffset,
+    },
+    ref
+  ) => {
+    const menuRef = useRef<HTMLDivElement>(null)
+
+    const handleUpdateStyles = (): void => {
+      if (!triggerRef.current || !menuRef.current) {
+        return
+      }
+
+      const menuStyles = {
+        ...calculateRightClickMenuStyles(mouseOffset, triggerRef, menuRef),
+        ...style,
+      }
+
+      const menuStyleString = convertCSSPropertiesToString(menuStyles)
+
+      if (menuRef.current) {
+        menuRef.current.setAttribute('style', menuStyleString)
+      }
+    }
+
+    const rightClickMenuClassName = classnames('cf-right-click', {
+      [`${className}`]: className,
+      [`cf-right-click__${color}`]: color,
+    })
+
+    const hidePopoverWhenOutOfView = (
+      entries: IntersectionObserverEntry[]
+    ): void => {
+      if (!!entries.length && entries[0].isIntersecting === false) {
+        onHide()
+      }
+    }
+
+    const observer = new IntersectionObserver(hidePopoverWhenOutOfView)
+
+    useLayoutEffect((): (() => void) => {
+      handleUpdateStyles()
+      observer.observe(triggerRef.current)
+      // The third argument in addEventListener is "false" by default and controls bubbling
```

Comment:
😆 and no one remembers why until this moment

Reply:
Helpful comment is helpful